### ISSUE TITLE
Creating new branch from HEAD should switch silently

### ISSUE THREAD

**Issue by rebornix** [microsoft/vscode-remote-repositories-github#227]
* Open a repository in vscode.dev
* Create a new branch and switch to that branch from the quick pick
* ❓Instead of switching silently (like desktop), it will reload the window

**Comment by joyceerhl** [comment]
I believe this is blocked on https://github.com/microsoft/vscode/issues/35109 (cc @bpasero). RemoteHub URIs encode information about the current branch/tag/commit in the authority, and since workspace ids are computed as a hash of the uri, this means that every branch/tag/commit is a fundamentally new workspace. We cannot currently change workspace folders without requiring a full reload.

I agree that avoiding a reload would be nice here. FWIW, RemoteHub is also interested in silently changing the workspace without reloading for other scenarios, like to support the case in vscode.dev where we need to disambiguate the branch part of vscode.dev url for a ref containing slashes in a private repo--currently the embedder contains layer breaking `ICredentialsProvider` code to enable authenticated octokit calls to validate the branch ref. If we had the ability to switch folders without a reload, we would simply be able to delegate validation of the workspace uri to RemoteHub and silently swap out the workspace uri post-validation, rather than having to get it right at the time of workbench creation.

**Comment by bpasero** [comment]
We have a similar problem for the local file system support on vscode.dev where we actually cannot tolerate the window reload after opening a folder because that is just not possible with local file system access. So what we do is actually always open an empty workspace (`.code-workspace`) and fake every "Open Folder" call to actually be a "Add Folder to Workspace" which works without window reload.

So in essence, the workspace does not change, only the folders within. But it is really just a workaround for not having https://github.com/microsoft/vscode/issues/35109

**Comment by joyceerhl** [comment]
@bpasero Ideally we would not need a workaround, do we have any blockers on https://github.com/microsoft/vscode/issues/35109? If that is unlikely to happen any time soon, are there any disadvantages for the empty workspace approach we use for local fs support? I wonder if we could use that same workaround to support switching branches without a reload in github.dev and vscode.dev.

**Comment by bpasero** [comment]
No blockers but also not just 1 day of work, see https://github.com/microsoft/vscode/issues/35109#issuecomment-817905964

No downsides from the workspace approach I would say, but it pretty much disables real multi-folder workspace support, not sure we have that in use today.

**Comment by joyceerhl** [comment]
> disables real multi-folder workspace support

I was just chatting with a Codespaces dev about the work they're doing to implement multi-root folder support in github.dev yesterday--it seems the way that web embedders would support multi root workspaces is by having the embedder accept a query param pointing to the workspace file URI, setting `IWorkspaceProvider.workspace.workspaceUri` to the value of that URI, and letting VS Code resolve the workspace file that way. So could we have the following scenarios:
* When there is a workspace file URI specified in embedder query parameters, the embedder will just use that workspace file
* Otherwise (e.g. for single workspace folder case), we create an initially empty workspace as you described and add the workspace folder to it
* In both cases, changing branches e.g. from `main` to `dev/foo` would remove the workspace folder at `vscode-vfs://github/microsoft/vscode` and add the workspace folder at `vscode-vfs://github%2B7b2276223a312c22726566223a7b2274797065223a/microsoft/vscode` () to represent a branch change

If that is a workable approach I think it would help @dbaeumer's multi-root GitHub remote workspace use case for CPython + WASM as well, where we want to build a multi-root workspace comprising entirely `vscode-vfs` workspace folder URIs.

**Comment by bpasero** [comment]
Yes, it is entirely up to the embedder to provide a multi root workspace URI. I had commented this back to the Codespace team previously.

If you run `scripts/code-web.sh` you can actually see MR support in action from our in memory embedder.

Btw Codespaces supports MR workspaces when you have a container backend, just not without.

**Comment by joyceerhl** [comment]
Closing as duplicate of https://github.com/microsoft/vscode-remote-repositories-github/issues/255
